File: LeetCode/Jump Game II.cpp
#include <algorithm>
#include <iostream>
using namespace std;
class Solution {
public:
    int jump(int A[], int n) {
        int idx = 0, level = 0, maxLength = 0;
        if (n == 1) return 0;
        while (idx < n){
            level++;
            if (level >= n) return 0;
            int i = idx, range = min(maxLength, n-1);
            while (i <= range){
                maxLength = max(maxLength, i + A[i]);
                if (maxLength >= n-1)
                    return level;
                i++;
            }
            idx = i;
        }
        return 0;
    }
};
int main(int argc, char const *argv[])
{
    Solution sol;
    int A[]= {3,4,3,1,0,0,0,3,0,2,0,3}, n = 12;
    cout << sol.jump(A, n) << endl;
    return 0;
}

File: LeetCode/Jump Game II.py
class Solution:
    # @param A, a list of integers
    # @return an integer
    def jump(self, A):
        if len(A)==1:
            return 0

        step = 0
        end = A[0]
        start = 0
        while end < len(A)-1 and start < end:
            step+=1
            tmp = end
            for i in range(start+1 , end+1):
                if (A[i]+i > end):
                    end = A[i] + i
            start = tmp

        return step+1

if __name__ == '__main__':
    s = Solution()
    # A = [3,2,1,0,4]
    A = [2,3,1,2,4]
    A = [2,2,3,2,4]
    A = [2,1]
    # A = [1,1,1,1]
    print s.jump(A)